ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -0,0 +1,775 @@
+	uint8_t bbool = 0;
+	uint32_t size = sizeof(uint8_t);
+	void *attr = NULL;
+
+	rc = get_attribute(templ, attribute, &bbool, &size);
+	if (rc) {
+		if (rc != PKCS11_RV_NOT_FOUND)
+			return rc;
+		attr = pkcs11_object_default_boolprop(attribute);
+		if (!attr)
+			return PKCS11_CKR_TEMPLATE_INCOMPLETE;
+	} else {
+		attr = &bbool;
+	}
+
+	/* Boolean attributes are 1byte in the ABI, no alignment issue */
+	return add_attribute(out, attribute, attr, sizeof(uint8_t));
+}
+
+static enum pkcs11_rc set_mandatory_boolprops(struct obj_attrs **out,
+					      struct obj_attrs *temp,
+					      uint32_t const *bp,
+					      size_t bp_count)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	size_t n = 0;
+
+	for (n = 0; n < bp_count; n++) {
+		rc = pkcs11_import_object_boolprop(out, temp, bp[n]);
+		if (rc)
+			return rc;
+	}
+
+	return rc;
+}
+
+static enum pkcs11_rc set_mandatory_attributes(struct obj_attrs **out,
+					       struct obj_attrs *temp,
+					       uint32_t const *bp,
+					       size_t bp_count)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	size_t n = 0;
+
+	for (n = 0; n < bp_count; n++) {
+		uint32_t size = 0;
+		void *value = NULL;
+
+		if (get_attribute_ptr(temp, bp[n], &value, &size))
+			return PKCS11_CKR_TEMPLATE_INCOMPLETE;
+
+		rc = add_attribute(out, bp[n], value, size);
+		if (rc)
+			return rc;
+	}
+
+	return rc;
+}
+
+static enum pkcs11_rc get_default_value(enum pkcs11_attr_id id, void **value,
+					uint32_t *size)
+{
+	/* should have been taken care of already */
+	assert(!pkcs11_attr_is_boolean(id));
+
+	if (id == PKCS11_CKA_PUBLIC_KEY_INFO) {
+		EMSG("Cannot provide default PUBLIC_KEY_INFO");
+		return PKCS11_CKR_TEMPLATE_INCONSISTENT;
+	}
+
+	/* All other attributes have an empty default value */
+	*value = NULL;
+	*size = 0;
+	return PKCS11_CKR_OK;
+}
+
+static enum pkcs11_rc set_optional_attributes(struct obj_attrs **out,
+					      struct obj_attrs *temp,
+					      uint32_t const *bp,
+					      size_t bp_count)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	size_t n = 0;
+
+	for (n = 0; n < bp_count; n++) {
+		uint32_t size = 0;
+		void *value = NULL;
+
+		rc = get_attribute_ptr(temp, bp[n], &value, &size);
+		if (rc == PKCS11_RV_NOT_FOUND)
+			rc = get_default_value(bp[n], &value, &size);
+		if (rc)
+			return rc;
+
+		rc = add_attribute(out, bp[n], value, size);
+		if (rc)
+			return rc;
+	}
+
+	return rc;
+}
+
+/*
+ * Below are listed the mandated or optional expected attributes for
+ * PKCS#11 storage objects.
+ *
+ * Note: boolprops (mandated boolean attributes) PKCS11_CKA_ALWAYS_SENSITIVE,
+ * and PKCS11_CKA_NEVER_EXTRACTABLE are set by the token, not provided
+ * in the client template.
+ */
+
+/* PKCS#11 specification for any object (session/token) of the storage */
+static const uint32_t pkcs11_any_object_boolprops[] = {
+	PKCS11_CKA_TOKEN, PKCS11_CKA_PRIVATE,
+	PKCS11_CKA_MODIFIABLE, PKCS11_CKA_COPYABLE, PKCS11_CKA_DESTROYABLE,
+};
+
+static const uint32_t pkcs11_any_object_optional[] = {
+	PKCS11_CKA_LABEL,
+};
```

Comment:
@etienne-lms it is unclear to me if label is option or mandated but default to empty. CKA_LABEL is also the only attribute that can be changed after the object was created, so it should be fine to start with an empty value as then the logic to handle the updates might become easier.